theano/gof/wrapper.py

```diff
@@ -0,0 +1,419 @@
+
+    If your op contains props `attr1` AND `attr2`, the op.get_params() method will
+    automatically try to look for it and generate an appropriate wrapped struct.
+    The props must be able to pass the filtering (not strict, downcasting allowed)
+    of corresponding types defined into Wrapper.
+
+        __props__ = ('attr1', 'attr2')
+        def __init__(value_attr1, value_attr2):
+            self.attr1 = value_attr1
+            self.attr2 = value_attr2
+
+    In perform() implementation (with params named `param`):
+
+        var1 = param.attr1
+        var2 = param.attr2
+
+    In c_code() implementation (with `param = sub['params']`):
+
+        PyArrayObject* attr1 = param.attr1;
+        PyArrayObject* attr2 = param.attr2;
+        /* You won't need to free them or whatever else. */
+
+
+See `theano/gof/tests/test_wrapper.py` for a complete working example.
+
+"""
+
+from __future__ import absolute_import, print_function, division
+import re
+import hashlib
+import numpy
+from theano.gof.utils import MethodNotDefined
+from theano.gof import Type
+from theano.tensor.utils import hash_from_ndarray
+
+# NB: Maybe we should check if an attribute name is a C/C++ keyword, and raise an error if so.
+# These are some lists of C/C++ keywords:
+# http://fr.cppreference.com/w/cpp/keyword
+# http://fr.cppreference.com/w/c/keyword
+
+
+class Wrap(dict):
+    """
+    Internal convenient class to wrap many Python objects into one
+    (this class is not safe as the hash method does not check if values are effectively hashable).
+
+    Example:
+        >>> w = Wrap(attr1=1, attr2=2.0, attri='3')
+        >>> print(w.attr1, w.attr2, w.attri)
+        >>> d = dict(a=1, b=2, c='test')
+        >>> w2 = Wrap(**d)
+        >>> print(w2.a, w2.b, w2.c)
+
+    """
+
+    def __init__(self, **kwargs):
+        super(Wrap, self).__init__(**kwargs)
+        if len(kwargs) == 0:
+            raise TypeError('Wrap: cannot wrap empty data.')
+
+    def __repr__(self):
+        return 'Wrap(%s)' % ', '.join([('%s:%s' % (k, type(self[k]))) for k in sorted(self.keys())])
+
+    def __getattr__(self, key):
+        if key not in self:
+            raise AttributeError('Wrap: attribute "%s" does not exist.' % key)
+        return self[key]
+
+    def __hash__(self):
+        keys = sorted(self.keys())
+        types = []
+        attributes = []
+        for k in keys:
+            types += (type(self[k]),)
+            if isinstance(self[k], numpy.ndarray):
+                # Note: hash_from_ndarray returns a string, so the hash is not yet complete
+                # (__hash__ must return an integer).
+                attributes += (hash_from_ndarray(self[k]),)
+            else:
+                # No checking, data should be hashable.
+                attributes += (self[k],)
+        return hash((type(self),) + tuple(keys) + tuple(types) + tuple(attributes))
```

Comment:
As you inherit from dict and implement \__hash__, can you  override the methods that will make sure people won't modify the content of the dict? I mean, don't add new keys, don't remove keys, don't change the values of keys.